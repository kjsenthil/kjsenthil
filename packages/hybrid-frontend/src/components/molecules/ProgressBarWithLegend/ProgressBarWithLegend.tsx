import * as React from 'react';
import { ProgressBar, Spacer } from '../../atoms';
import { ProgressLegendContainer, ProgressLegendWrapper } from './ProgressBarWithLegend.styles';
import Legend, { LegendProps } from '../Legend';

interface ProgressBarData {
  progress: number;
  legendProps: Pick<LegendProps, 'title' | 'value'>;
}

export interface ProgressBarWithLegendProps {
  currencyFormatter: (num: number) => string;
  progressBarData: Array<ProgressBarData>;
}

export default function ProgressBarWithLegend({
  currencyFormatter,
  progressBarData,
}: ProgressBarWithLegendProps) {
  const progress = progressBarData.map((data) => data.progress);

  return (
    <>
      <ProgressBar height={38} borderRadius={6} progress={progress} />
      <Spacer y={2} />
      <ProgressLegendContainer>
        {progressBarData.map((data) => {
          const width = data.progress * 100;
          return (
            <ProgressLegendWrapper
              width={width}
              key={`${data.progress}-${data.legendProps.title}-${data.legendProps.value}`}
            >
              <Legend {...data.legendProps} valueFormatter={currencyFormatter} />
            </ProgressLegendWrapper>
          );
        })}
      </ProgressLegendContainer>
    </>
  );
}
