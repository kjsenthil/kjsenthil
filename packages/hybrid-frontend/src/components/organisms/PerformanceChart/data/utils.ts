export interface GetPerformanceDataResponse {
  // All numbers have 2 decimal places
  Data: {
    Type: 'performance';
    Id: string;
    Attributes: {
      AccountValue: number;
      Performance: {
        Value: number;
        Percentage: number; // This is a float. 50% will be 50.00
      };
      Values: Array<{ Date: string; Value: number }>;
    };
    Links: {
      self: string;
    };
    Relationships: {
      contributions: {
        Links: {
          related: string;
        };
        Data: {
          Type: 'contributions';
          Id: string;
        };
      };
    };
  };
  Included: Array<{
    // All numbers have 3 decimal places
    Type: 'contributions';
    Id: string;
    Attributes: {
      TotalContributions: number;
      Contributions: Array<{ Date: string; NetContributionsToDate: number }>;
    };
    Links: {
      self: string;
    };
    Relationships: null;
  }>;
}

export interface TimeSeriesDatum {
  date: Date;
  value: number;
}

export interface PerformanceDatum extends TimeSeriesDatum {
  metadata?: Record<string, unknown>;
}

export interface ContributionDatum extends TimeSeriesDatum {
  metadata?: Record<string, unknown>;
}

export enum PerformanceChartPeriod {
  '1M' = '1m',
  '3M' = '3m',
  '6M' = '6m',
  '1Y' = '1y',
  'ALL_TIME' = 'alltime',
}

export function sortTimeSeries<TS extends TimeSeriesDatum>(
  timeSeriesData: TS[],
  direction: 'asc' | 'desc' = 'asc'
): TS[] {
  return [...timeSeriesData].sort((d1, d2) =>
    direction === 'asc'
      ? d2.date.getTime() - d1.date.getTime()
      : d1.date.getTime() - d2.date.getTime()
  );
}

export function getTimeSeriesMinMax<TS extends TimeSeriesDatum>(timeSeriesData: TS[]) {
  if (timeSeriesData.length === 0) {
    return {
      minDate: new Date(0),
      maxDate: new Date(0),
      minValue: 0,
      maxValue: 0,
    };
  }

  let minDate = timeSeriesData[0].date;
  let maxDate = timeSeriesData[0].date;
  let minValue = timeSeriesData[0].value;
  let maxValue = timeSeriesData[0].value;

  timeSeriesData.forEach(({ date, value }) => {
    if (date > maxDate) maxDate = date;
    if (date < minDate) minDate = date;
    if (value > maxValue) maxValue = value;
    if (value < minValue) minValue = value;
  });

  return {
    minDate,
    maxDate,
    minValue,
    maxValue,
  };
}

export function timeSeriesDateAccessor(d: TimeSeriesDatum) {
  return d.date;
}

export function timeSeriesValueAccessor(d: TimeSeriesDatum) {
  return d.value;
}
