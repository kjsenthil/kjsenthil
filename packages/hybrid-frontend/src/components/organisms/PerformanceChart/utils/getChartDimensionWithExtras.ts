import { ChartDimension, ChartDimensionWithExtras } from '../../../../config/chart';

/**
 * This function calculates and injects certain extra figures (like inner width
 * and inner height) to a chart dimension object. It returns a new chart
 * dimension object containing these extras.
 */
export default function getChartDimensionWithExtras(
  chartDimension: ChartDimension
): ChartDimensionWithExtras {
  return {
    ...chartDimension,
    innerWidth: chartDimension.width - chartDimension.margin.left - chartDimension.margin.right,
    innerHeight: chartDimension.height - chartDimension.margin.top - chartDimension.margin.bottom,
  };
}
