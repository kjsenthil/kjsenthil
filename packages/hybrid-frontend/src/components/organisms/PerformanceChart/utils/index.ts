/* eslint-disable import/prefer-default-export */
export { default as getChartDimensionWithExtras } from './getChartDimensionWithExtras';
