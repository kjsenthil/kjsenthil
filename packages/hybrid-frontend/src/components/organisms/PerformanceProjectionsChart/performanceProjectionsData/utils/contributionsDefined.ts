import { ProjectionsChartProjectionDatum } from '../../../../../services/projections';

/**
 * This is a d3-style "defined" function. SVG paths are only drawn when this
 * function returns true. Used when we want to "hide" certain bits of a graph,
 * for example.
 */
export default function contributionsDefined(d: ProjectionsChartProjectionDatum): boolean {
  return d.netContributionsToDate > 0;
}
