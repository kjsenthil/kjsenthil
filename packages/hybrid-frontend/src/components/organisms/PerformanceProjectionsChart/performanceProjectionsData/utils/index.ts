export { default as valueDefinedFactory } from './contributionsDefined';
export { default as mapDate } from './mapDate';

export * from './accessors';
export * from './minMax';
