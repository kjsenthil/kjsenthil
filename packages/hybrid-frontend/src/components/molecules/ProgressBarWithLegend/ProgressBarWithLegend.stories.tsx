import React from 'react';
import { Meta, Story } from '@storybook/react/types-6-0';
import ProgressBarWithLegend, { ProgressBarWithLegendProps } from '.';
import { formatCurrency } from '../../../utils/formatters';

export default {
  title: 'Digital Hybrid/Molecules/ProgressBar With Legend',
  component: ProgressBarWithLegend,
  argTypes: {},
} as Meta;

const progressBarData = [
  {
    progress: 0.2,
    legendProps: { title: 'Lump Sum', value: 20000 },
  },
  {
    progress: 0.1,
    legendProps: { title: 'From Age 67 - 89', value: 1080000 },
  },
  {
    progress: 0.05,
    legendProps: { title: 'Remaining', value: 72000 },
  },
];

const Template: Story<ProgressBarWithLegendProps> = (args) => <ProgressBarWithLegend {...args} />;

export const Default = Template.bind({});
Default.args = {
  currencyFormatter: formatCurrency,
  progressBarData,
};
