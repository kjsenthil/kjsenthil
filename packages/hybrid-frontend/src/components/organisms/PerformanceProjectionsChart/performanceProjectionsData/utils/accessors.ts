import { TimeSeriesDatum } from '../../../../../utils/data';
import {
  ProjectionsChartProjectionDatum,
  ProjectionsChartProjectionTargetDatum,
} from '../../../../../services/projections';

export function dateAccessor(d: TimeSeriesDatum): Date {
  return d.date;
}

export function upperBoundAccessor(d: ProjectionsChartProjectionDatum): number {
  return d.upperBound;
}

export function lowerBoundAccessor(d: ProjectionsChartProjectionDatum): number {
  return d.lowerBound;
}

export function valueAccessor(d: ProjectionsChartProjectionDatum): number {
  return d.value;
}

export function contributionsAccessor(d: ProjectionsChartProjectionDatum): number {
  return d.netContributionsToDate;
}

export function valueTargetAccessor(d: ProjectionsChartProjectionTargetDatum): number {
  return d.value;
}
