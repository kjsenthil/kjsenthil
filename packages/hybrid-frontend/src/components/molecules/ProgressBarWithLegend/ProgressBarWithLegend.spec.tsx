import * as React from 'react';
import { renderWithTheme } from '@tsw/test-util';
import { formatCurrency } from '../../../utils/formatters';
import ProgressBarWithLegend from './ProgressBarWithLegend';

const progressBarData = [
  {
    progress: 0.2,
    legendProps: { title: 'Lump Sum', value: 20000 },
  },
  {
    progress: 0.5,
    legendProps: { title: 'From Age 67 - 89', value: 1080000 },
  },
  {
    progress: 0.05,
    legendProps: { title: 'Remaining', value: 72000 },
  },
];

const props = {
  currencyFormatter: formatCurrency,
  progressBarData,
};

describe('PrgoressBarWithLegend', () => {
  it('matches snapshot', () => {
    expect(renderWithTheme(<ProgressBarWithLegend {...props} />)).toMatchSnapshot();
  });
});
