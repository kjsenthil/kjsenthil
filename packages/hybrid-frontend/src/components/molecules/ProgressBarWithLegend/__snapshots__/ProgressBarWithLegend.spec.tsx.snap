// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrgoressBarWithLegend matches snapshot 1`] = `
Object {
  "result": Object {
    "asFragment": [Function],
    "baseElement": .c0 {
  position: relative;
  width: 100%;
  height: 2.375rem;
  border-radius: 6px;
  overflow: hidden;
  background: #fff;
}

.c2 {
  position: absolute;
  top: 0;
  left: 0;
  height: 2.375rem;
  width: 75%;
  border-radius: 6px;
  background: #a5f2ec;
  -webkit-animation: jbfzvH 1s ease-out;
  animation: jbfzvH 1s ease-out;
  z-index: 1;
}

.c3 {
  position: absolute;
  top: 0;
  left: 0;
  height: 2.375rem;
  width: 70%;
  border-radius: 6px;
  background: linear-gradient(to left,#00e8d9,#00c4bd 50%);
  -webkit-animation: jZJhoi 1s ease-out;
  animation: jZJhoi 1s ease-out;
  z-index: 1;
}

.c4 {
  position: absolute;
  top: 0;
  left: 0;
  height: 2.375rem;
  width: 20%;
  border-radius: 6px;
  background: #009492;
  -webkit-animation: gDKouL 1s ease-out;
  animation: gDKouL 1s ease-out;
  z-index: 1;
}

.c1 {
  position: absolute;
  top: 0;
  left: 0;
  height: 100%;
  width: 100%;
  border-radius: 6px;
  border: 2px dashed #ebb300;
}

.c8 {
  font-size: 0.75rem;
  line-height: 1rem;
  -webkit-letter-spacing: 0.01875rem;
  -moz-letter-spacing: 0.01875rem;
  -ms-letter-spacing: 0.01875rem;
  letter-spacing: 0.01875rem;
  font-weight: bold;
  color: #6a6775;
  white-space: nowrap;
}

.c10 {
  font-size: 0.75rem;
  line-height: 1rem;
  -webkit-letter-spacing: 0.01875rem;
  -moz-letter-spacing: 0.01875rem;
  -ms-letter-spacing: 0.01875rem;
  letter-spacing: 0.01875rem;
  font-weight: bold;
  color: #380c57;
  white-space: nowrap;
}

.c5 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  width: 100%;
  overflow: hidden;
}

.c6 {
  min-width: 110px;
  max-width: 50%;
  width: 20%;
  -webkit-animation: gDKouL 1s ease-out;
  animation: gDKouL 1s ease-out;
  z-index: 1;
}

.c11 {
  min-width: 110px;
  max-width: 50%;
  width: 50%;
  -webkit-animation: bDhWlw 1s ease-out;
  animation: bDhWlw 1s ease-out;
  z-index: 1;
}

.c12 {
  min-width: 110px;
  max-width: 50%;
  width: 5%;
  -webkit-animation: dvwyrU 1s ease-out;
  animation: dvwyrU 1s ease-out;
  z-index: 1;
}

.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 8px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

<body>
      <div>
        <div
          aria-valuenow="75"
          class="c0"
          role="progressbar"
        >
          <div
            class="c1"
          />
          <div
            class="c2"
          />
          <div
            class="c3"
          />
          <div
            class="c4"
          />
        </div>
        <div
          class="MuiBox-root MuiBox-root-1"
          data-testid="Spacer"
        />
        <div
          class="c5"
        >
          <div
            class="c6"
            width="20"
          >
            <div
              class="c7"
            >
              <div>
                <h6
                  class="MuiTypography-root c8 MuiTypography-body1"
                >
                  LUMP SUM
                </h6>
                <div
                  class="c9"
                >
                  <h6
                    class="MuiTypography-root c10 MuiTypography-body1"
                  >
                    £20,000.00
                  </h6>
                </div>
              </div>
            </div>
          </div>
          <div
            class="c11"
            width="50"
          >
            <div
              class="c7"
            >
              <div>
                <h6
                  class="MuiTypography-root c8 MuiTypography-body1"
                >
                  FROM AGE 67 - 89
                </h6>
                <div
                  class="c9"
                >
                  <h6
                    class="MuiTypography-root c10 MuiTypography-body1"
                  >
                    £1,080,000.00
                  </h6>
                </div>
              </div>
            </div>
          </div>
          <div
            class="c12"
            width="5"
          >
            <div
              class="c7"
            >
              <div>
                <h6
                  class="MuiTypography-root c8 MuiTypography-body1"
                >
                  REMAINING
                </h6>
                <div
                  class="c9"
                >
                  <h6
                    class="MuiTypography-root c10 MuiTypography-body1"
                  >
                    £72,000.00
                  </h6>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </body>,
    "container": <div>
      <div
        aria-valuenow="75"
        class="sc-iCfLBT fLrhme"
        role="progressbar"
      >
        <div
          class="sc-pVTma kWowoK"
        />
        <div
          class="sc-furvIG eUDfdF"
        />
        <div
          class="sc-furvIG kDqjiJ"
        />
        <div
          class="sc-furvIG jaMmSV"
        />
      </div>
      <div
        class="MuiBox-root MuiBox-root-1"
        data-testid="Spacer"
      />
      <div
        class="sc-kfPtPH bEVDNH"
      >
        <div
          class="sc-fKVsgm kGJYoa"
          width="20"
        >
          <div
            class="sc-bBHwJV kXmoUd"
          >
            <div>
              <h6
                class="MuiTypography-root sc-jrQzUz EJbHG MuiTypography-body1"
              >
                LUMP SUM
              </h6>
              <div
                class="sc-iwjezw hwFGdy"
              >
                <h6
                  class="MuiTypography-root sc-jrQzUz prRFo MuiTypography-body1"
                >
                  £20,000.00
                </h6>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-fKVsgm elpSEl"
          width="50"
        >
          <div
            class="sc-bBHwJV kXmoUd"
          >
            <div>
              <h6
                class="MuiTypography-root sc-jrQzUz EJbHG MuiTypography-body1"
              >
                FROM AGE 67 - 89
              </h6>
              <div
                class="sc-iwjezw hwFGdy"
              >
                <h6
                  class="MuiTypography-root sc-jrQzUz prRFo MuiTypography-body1"
                >
                  £1,080,000.00
                </h6>
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-fKVsgm eGrvXg"
          width="5"
        >
          <div
            class="sc-bBHwJV kXmoUd"
          >
            <div>
              <h6
                class="MuiTypography-root sc-jrQzUz EJbHG MuiTypography-body1"
              >
                REMAINING
              </h6>
              <div
                class="sc-iwjezw hwFGdy"
              >
                <h6
                  class="MuiTypography-root sc-jrQzUz prRFo MuiTypography-body1"
                >
                  £72,000.00
                </h6>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>,
    "debug": [Function],
    "findAllByAltText": [Function],
    "findAllByDisplayValue": [Function],
    "findAllByLabelText": [Function],
    "findAllByPlaceholderText": [Function],
    "findAllByRole": [Function],
    "findAllByTestId": [Function],
    "findAllByText": [Function],
    "findAllByTitle": [Function],
    "findByAltText": [Function],
    "findByDisplayValue": [Function],
    "findByLabelText": [Function],
    "findByPlaceholderText": [Function],
    "findByRole": [Function],
    "findByTestId": [Function],
    "findByText": [Function],
    "findByTitle": [Function],
    "getAllByAltText": [Function],
    "getAllByDisplayValue": [Function],
    "getAllByLabelText": [Function],
    "getAllByPlaceholderText": [Function],
    "getAllByRole": [Function],
    "getAllByTestId": [Function],
    "getAllByText": [Function],
    "getAllByTitle": [Function],
    "getByAltText": [Function],
    "getByDisplayValue": [Function],
    "getByLabelText": [Function],
    "getByPlaceholderText": [Function],
    "getByRole": [Function],
    "getByTestId": [Function],
    "getByText": [Function],
    "getByTitle": [Function],
    "queryAllByAltText": [Function],
    "queryAllByDisplayValue": [Function],
    "queryAllByLabelText": [Function],
    "queryAllByPlaceholderText": [Function],
    "queryAllByRole": [Function],
    "queryAllByTestId": [Function],
    "queryAllByText": [Function],
    "queryAllByTitle": [Function],
    "queryByAltText": [Function],
    "queryByDisplayValue": [Function],
    "queryByLabelText": [Function],
    "queryByPlaceholderText": [Function],
    "queryByRole": [Function],
    "queryByTestId": [Function],
    "queryByText": [Function],
    "queryByTitle": [Function],
    "rerender": [Function],
    "unmount": [Function],
  },
}
`;
